### PR TITLE
[CephAdm][Automation]Bootstrap with spec file

### PR DESCRIPTION
**Pull request by MohitBis**
[CephAdm][Automation]Bootstrap with spec file

Scenario: Bootstrap a cluster along with a spec file using "--apply-spec"

### CODE REVIEW

**Comment by MohitBis** [comment]
Logs:
quincy: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-UT0REH/

Reef: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-NL4P6D/

File: tests/cephadm/test_cephadm_bootstrap_with_spec_file.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    get_tools_repo,
+    setup_client_node,
+    setup_installer_node,
+)
+from utility.log import Log
+
+log = Log(__name__)
+
+
+def create_spec_file(installer, nodes, specs, spec_file):
```
Comment (vamahaja) [inline tests/cephadm/test_cephadm_bootstrap_with_spec_file.py:15]: Move this method here - https://github.com/red-hat-storage/cephci/blob/master/cli/ops/cephadm.py#L23

and make this part of https://github.com/red-hat-storage/cephci/blob/master/cli/ops/cephadm.py#L24

File: tests/cephadm/test_cephadm_bootstrap_with_spec_file.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    # Configure installer node
+    setup_installer_node(
+        installer, nodes, rhbuild, tools_repo, build_type, ibm_build, ansible_preflight
+    )
+
+    # Get bootstrap config
+    bootstrap_config = config.get("bootstrap")
+
+    # create spec file
+    create_spec_file(
```
Comment (vamahaja) [inline tests/cephadm/test_cephadm_bootstrap_with_spec_file.py:75]: as this is part of bootstrap, move this method to - https://github.com/red-hat-storage/cephci/blob/master/cli/ops/cephadm.py#L141

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: suites/reef/cephadm/tier-2-bootstrap-along-with-spec-file.yaml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+      polarion-id: CEPH-83575604
+      module: test_cephadm_bootstrap_with_spec_file.py
+      config:
+        ansible_preflight: true
+        bootstrap:
+          mon-ip: node1
+          apply-spec: specs.yaml
+        client:
+          ansible_clients: true
+        specs:
```
Comment (vamahaja) [inline suites/reef/cephadm/tier-2-bootstrap-along-with-spec-file.yaml:27]: This specs should be under `bootstrap` key

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: tests/cephadm/test_cephadm_bootstrap_with_spec_file.py (hunk 1 above)
Comment (MohitBis) [reply tests/cephadm/test_cephadm_bootstrap_with_spec_file.py]: Thanks @vamahaja 
Updated

File: suites/reef/cephadm/tier-2-bootstrap-along-with-spec-file.yaml (hunk 3 above)
Comment (MohitBis) [reply suites/reef/cephadm/tier-2-bootstrap-along-with-spec-file.yaml]: Thanks @vamahaja
Updated

**Comment by MohitBis** [comment]
**Latest Logs attached.**
[Uploading Bootstrap_a_cluster_along_with_a_spec_file_0.log…]()


Note:
Not shared magan002 link since magna002 log path is unaccessable.

File: cli/ops/cephadm.py [hunk 4]
```diff
@@ -48,6 +50,19 @@ def generate_bootstrap_config(node, config):
         elif file_type == "json":
             json.dump(config, _f)
 
+        elif file_type == "yaml":
+            for spec in config["spec"]:
+                if spec["service_type"] == "host":
+                    hostname = spec["hostname"][-1]
+                    spec["addr"] = nodes[int(hostname) - 1].ip_address
```
Comment (vamahaja) [inline cli/ops/cephadm.py:57]: What we are expecting for hostname here ?

File: cli/ops/cephadm.py [hunk 5]
```diff
@@ -48,6 +50,19 @@ def generate_bootstrap_config(node, config):
         elif file_type == "json":
             json.dump(config, _f)
 
+        elif file_type == "yaml":
+            for spec in config["spec"]:
```
Comment (vamahaja) [inline cli/ops/cephadm.py:54]: Create separate method for this one and move call it inside bootstrap config parsing.

File: cli/ops/cephadm.py [hunk 6]
```diff
@@ -142,7 +157,7 @@ def bootstrap(
     # Check for bootstrap configs
     for k in kw.keys():
         if isinstance(kw.get(k), dict):
-            kw[k] = generate_bootstrap_config(installer, kw.get(k))
+            kw[k] = generate_bootstrap_config(installer, nodes, kw.get(k))
```
Comment (vamahaja) [inline cli/ops/cephadm.py:160]: Add check for `apply-spec` key and call method to generate config.

File: cli/ops/cephadm.py [hunk 7]
```diff
@@ -48,6 +50,19 @@ def generate_bootstrap_config(node, config):
         elif file_type == "json":
             json.dump(config, _f)
 
+        elif file_type == "yaml":
```
Comment (vamahaja) [inline cli/ops/cephadm.py:53]: expect only config and write data here.

File: tests/cephadm/test_cephadm_bootstrap_with_spec_file.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
+from cli.exceptions import ConfigError, ResourceNotFoundError
```
Comment (vamahaja) [inline tests/cephadm/test_cephadm_bootstrap_with_spec_file.py:1]: Can't we use existing test case - https://github.com/red-hat-storage/cephci/blob/master/tests/cephadm/test_cephadm_bootstrap.py ?

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: cli/ops/cephadm.py (hunk 5 above)
Comment (MohitBis) [reply cli/ops/cephadm.py:72]: Updated

File: cli/ops/cephadm.py (hunk 6 above)
Comment (MohitBis) [reply cli/ops/cephadm.py]: Updated

File: cli/ops/cephadm.py (hunk 7 above)
Comment (MohitBis) [reply cli/ops/cephadm.py:71]: Updated

File: tests/cephadm/test_cephadm_bootstrap_with_spec_file.py (hunk 8 above)
Comment (MohitBis) [reply tests/cephadm/test_cephadm_bootstrap_with_spec_file.py]: Updated

**Comment by MohitBis** [comment]
Latest Logs: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-AZ2T8W/

**Review by vamahaja — COMMENTED** [review]
LGTM

**Review by vamahaja — APPROVED** [review]

**Pull request merged**